AzureTable/Microsoft.DataTransfer.AzureTable/Sink/TableAPIBulkSinkAdapter.cs

```diff
@@ -0,0 +1,172 @@
+    using Microsoft.DataTransfer.AzureTable.Utils;
+    using Microsoft.DataTransfer.Basics;
+    using Microsoft.DataTransfer.Extensibility;
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    internal sealed class TableAPIBulkSinkAdapter : IDataSinkAdapter
+    {
+        private const long maxLengthInBytesPerDocument = 2 * 1024 * 1024;
+
+        private string _connectionString;
+        private string _tableName;
+        private bool _overwrite;
+        private long _maxInputBufferSizeInBytes;
+        private int _throughput;
+        private int _maxLengthInBytesPerBatch;
+
+        private CloudTable cloudtable;
+        private ConcurrentDictionary<string, TableBatchOperation> dict;
+        private InputSizeTracker inputSizeTracker;
+        private BatchSizeTracker batchSizeTracker;
+
+        public int MaxDegreeOfParallelism
+        {
+            get { return 1; }
+        }
+
+        public TableAPIBulkSinkAdapter(string connectionString, string tableName, 
+            bool overwrite, long maxInputBufferSizeInBytes, int throughput, int batchSize)
+        {
+            _connectionString = connectionString;
+            _tableName = tableName;
+            _overwrite = overwrite;
+            _maxInputBufferSizeInBytes = maxInputBufferSizeInBytes;
+            _throughput = throughput;
+            _maxLengthInBytesPerBatch = batchSize;
+
+            CloudStorageAccount storageAccount = CloudStorageAccount.Parse(_connectionString);
+
+            TableConnectionPolicy connectionPolicy = new TableConnectionPolicy()
+            {
+                UseDirectMode = true,
+                UseTcpProtocol = true,
+            };
+
+            CloudTableClient tableClient = storageAccount.CreateCloudTableClient(connectionPolicy: connectionPolicy);
+            cloudtable = tableClient.GetTableReference(_tableName);
+        }
+
+        public async Task InitializeAsync(CancellationToken cancellation)
+        {
+            inputSizeTracker = new InputSizeTracker(maxLengthInBytesPerDocument, _maxInputBufferSizeInBytes);
+            batchSizeTracker = new BatchSizeTracker(_maxLengthInBytesPerBatch, inputSizeTracker);
+            dict = new ConcurrentDictionary<string, TableBatchOperation>();
+            await cloudtable.CreateIfNotExistsAsync(IndexingMode.Consistent, _throughput, cancellation);
+        }
+
+        public async Task WriteAsync(IDataItem dataItem, CancellationToken cancellation)
+        {
+            var item = GetITableEntityFromIDataItem(dataItem);
+
+            inputSizeTracker.Add(item);
+
+            if (!dict.ContainsKey(item.PartitionKey))
+            {
+                dict[item.PartitionKey] = new TableBatchOperation();
+            }
+
+            var batchOperation = dict[item.PartitionKey];
+            batchOperation.InsertOrReplace(item);
```

Comment:
Why replace?

Reply:
I just thought it best to cleanly overwrite everything to be on the safer side.
An insert would skip over all records that have a matching paritionkey and rowkey. On an import failure it would be best to insert again -- and skip over rows that already exist.
There is a slight chance that the schema could change between the first (failed/trial) import run and the next one. In that off-chance case an insert will produce wrong results.
I can change it to Insert if this is not an issue for GA.

Reply:
Spoke with Kiran. Will not fix in this PR.